### ISSUE TITLE
Renames embedded in conditions.

### ISSUE THREAD

**Issue by Blady-Com** [Ada-Rapporteur-Group/User-Community-Input#81]
Let's take a function, its returned result is used both for a condition and for a processing following this condition.
For example `Ada.Strings.Fixed.Index` in Unix path decomposition:
```
      while Index (Path, "/") > 0 loop
         Delete (Path, Path'First, Index (Path, "/")); -- some processing with position of "/"
      end loop;
```
The result is used for the condition of the while loop and processing inside the loop.

Another example, the `GNAT.OS_Lib.Read` function returns the number of characters actually read when reading a file:
```
      if Read (FD, Buffer'Address, Buffer'Length) < Buffer'Length then
         Put_Line
           ("warning only" & Natural'Image (Read (FD, Buffer'Address, Buffer'Length)) &
            " bytes read."); -- some processing with number of read bytes
      end if;
```

In the first example, calling Index twice has only a performance impact, but in the second example calling Read twice is erroneous. The solution to these two trivial examples is to declare a local variable to the subroutine.
On the other hand, if calling the function requires previously executing code, you must then use a `declare` statement.

To avoid this heaviness and improve readability, I propose to put a renamed fonction embedded in the condition.
First exemple:
```
      while (Pos renames Index (Path, "/")) > 0 loop
         Delete (Path, Path'First, Pos); -- some processing with position of "/"
      end loop;
```

Second exemple:
```
      if (Count renames Read (FD, Buffer'Address, Buffer'Length)) < Buffer'Length then
         Put_Line
           ("warning only" & Natural'Image (Count) &
            " bytes read."); -- some processing with number of read bytes
      end if;
```

Thus, the fonction is called only once and no need to local variable or declare statement.
This renaming could occur for any condition in `if` and `while` statement.
The scope of the renaming is the `if` or `while` statement itself.
(Note : it could be present in condition for `exit` statement but there is no interest.)

**Comment by Richard-Wai** [comment]
I'm struggling to understand what is "wrong" with the traditional approach of introducing a declare block

`
declare  Count: constant Natural := Read (FD, Buffer'Access, Buffer'Length);
begin if Count < Buffer'Length then
    Put_Line (".."); -- etc
end if; end;
`

What's more, in a normal rename we'd need to define the type of the new name, and in this proposal, we'd have to introduce something like the C++ 'auto' type here.

I'd even go as far as to argue that the proposed syntax would be more suggestive of a subprogram renaming than the renaming of an anonymous return object.

**Comment by ARG-Editor** [comment]
Umm, Richard, in Ada 2022 the type name is optional in a renames. So the syntax he is proposing is just that of an ordinary object renames (in an unconventional place).

Speaking for myself, I see a number of issues here:

(1) Not sure why this is restricted to just if statements and while loops. I could imagine such a syntax being useful in case statements, for loops (in the range or object after the in or of), and probably others. The design principle of keeping the language consistent suggests that it should be available as widely as possible.

(2) Introducing a new place where names can be declared means that one has to have new kinds of declarative regions, and thus new places where declarations can be hidden. I would worry that users might get very confused if a name declared in this way hides something else -- this is not a place where experienced Ada programmers are going to look for declarations. This gets worse if this syntax is allowed in the middle of expressions (which it ought to be) as it would increase the level of complication of resolving/understanding expressions. Imagine the new kinds of obfuscations that this would allow:

`            if Count > Count renames Index (Str, "Pattern") then
`
(Here the first Count is some object declared in an outer scope, and the second is declared here.) 

(3) I'm unconvinced that this saves enough over the declare block for the complication of intra-expression declarations. Perhaps a better solution (if one is needed at all) would be to allow an optional declarative part on other compound statements. That would eliminate the extra "begin" and "end" in Richard's example. That is, something like:

```
    declare
        Count renames Read (FD, Buffer'Access, Buffer'Length);
    if Count < Buffer'Length then
        Put_Line (".." & Natural'Image (Count)); -- etc
    end if;
```

Note that the renames syntax is already allowed, the only new thing here is allowing a declarative part on an if statement.

(4) Remember to keep in mind Ada's design goals and their consequences. I talked about those in an article on RR's website:
    http://www.rrsoftware.com/html/blog/consequences.html

                                  Randy.

**Comment by Richard-Wai** [comment]
> Umm, Richard, in Ada 2022 the type name is optional in a renames.

Yep I forgot about that!

>Perhaps a better solution (if one is needed at all) would be to allow an optional declarative part on other compound >statements. That would eliminate the extra "begin" and "end" in Richard's example. That is, something like:
> 
> ```
>     declare
>         Count renames Read (FD, Buffer'Access, Buffer'Length);
>     if Count < Buffer'Length then
>         Put_Line (".." & Natural'Image (Count)); -- etc
>     end if;
> ```
> 

This is actually really interesting!

**Comment by sttaft** [comment]
One approach is to simply eliminate the need for declare blocks, and like many other languages, permit at least some kinds of declarations pretty much anywhere a statement is permitted.  Declarations and statements are largely syntactically distinct from each other in Ada.

AdaCore has prototyped this in GNAT.  Here is a writeup of the prototyped feature:

    https://github.com/AdaCore/ada-spark-rfcs/blob/master/prototyped/rfc-local-vars-without-block.md

**Comment by Blady-Com** [comment]
> That is, something like:
> 
> ```
>     declare
>         Count renames Read (FD, Buffer'Access, Buffer'Length);
>     if Count < Buffer'Length then
>         Put_Line (".." & Natural'Image (Count)); -- etc
>     end if;
> ```
I actually like this too.
Let's not forget my first example which may be changed in:
```
      declare
         Pos renames Index (Path, "/");
      while Pos > 0 loop
         Delete (Path, Path'First, Pos); -- some processing with position of "/"
      end loop;
```
Of course, it assumes that the declare part is executed at each turn of the loop that is however less intuitive.

**Comment by CKWG** [comment]
### Ada - where art thou heading!

> https://github.com/AdaCore/ada-spark-rfcs/blob/master/prototyped/rfc-local-vars-without-block.md
The example from this link is a Trumpism, bad joke!
```
if X > 5 then
   X := X + 1;
   Squared : constant Integer := X**2;
   X := X + Squared;
else
   X := X - 1;
   Cubed : constant Integer := X**3;
   X := X + Cubed;
end if;
```
Blady's examples are more convincing.

However, that's nefandous - a declaration outside a loop but performed inside! **Poor Ada, where art thou heading!**
> ```
>       declare
>          Pos renames Index (Path, "/");
>       while Pos > 0 loop
>          Delete (Path, Path'First, Pos); -- some processing with position of "/"
>       end loop;
> ```
> Of course, it assumes that the declare part is executed at each turn of the loop that is however less intuitive.

I don't buy the argument that deeper indenting with declare blocks obfuscates the code. From the link:
> Clearly, the programmer can always fall back to using a local declare block for more complex requirements, though of
> course this is pushing more levels of indendatation on the more complex cases.
> Future possibilities ... allow arbitrary declarations

Keeping indentation small is just one art in programming of how to make code clearly arranged.

In summary: We've already had good solutions for this since Ada 83!
```
loop
  declare
    Pos renames Index (Path, "/");
  begin
    exit when Pos = 0;
    Delete (Path, Path'First, Pos); -- some processing with position of "/"
  end;
end loop;
```
Three more lines - and this is far better structured in my opinion.

**Comment by joshua-c-fletcher** [comment]
Scope is definitely less clear in this proposal, but things would persist until the first 'end' they hit.

The "Drawbacks" section states:

> Readability is one concern, but it is arguable that the current rules are worse from a readability point of view, as they tend to lead to overly indented source code, or perhaps even worse, to local variables being declared with a longer lifetime than they need, making it harder to understand the role the variable might be playing in the large amount of code where it is visible.

These late-declared items would only have shorter lifetimes when contrasted to an approach that avoids declare blocks and declares as much as possible at the top of a procedure (which certainly does happen sometimes).

Using declare blocks where they're appropriate would still allow better control of out-scoping locally declared items.  If the new proposal is adopted, there'll probably be a mix of declared items with shorter scope (for being declared later, or inside another  block), and other items having longer scope because no one bothered to limit it with a local declare block.

I can think of one statement that allows declaring something without a 'begin' and that is the extended return statement that's needed for limited type constructor functions.  They fairly uniquely declare a single variable for the scope of the extended return statement.  It's neat, but I can't think of anything else quite like it.

One approach that could reduce the need for full-on declare blocks inside other block constructs that effectively have the same scope would be to allow something like this:
(optional declarative region inside the loop, indicated with a 'declare' and a 'begin', but no extra 'end')
```
declare loop
   Pos renames Index (Path, "/");
begin
   exit when Pos = 0;
   Delete (Path, Path'First, Pos); -- some processing with position of "/"
end loop; 
```
```
declare while Pos > 0 loop
   Pos renames Index (Path, "/");
begin
   Delete (Path, Path'First, Pos); -- some processing with position of "/"
end loop;
```

This next one's a bit odd, and I took out the increment/decrement part, because those would have required the more typical declare blocks without other changes.... The idea is to use the blocks we already have, if the scope we need for the declarative region is the same.  This certainly looked much better with the loops than for this if/else block, though.

(same approach as for the loops - use 'declare' to indicate an optional declarative region inside the statement, with begin to end the declarative region, and no extra 'end')
```
declare if X > 5 then
   Squared : constant Integer := X**2;
begin
   X := @ + Squared;
declare else
   Cubed : constant Integer := X**3;
begin
   X := @ + Cubed;
end if;
```

I suppose this  isn't really to far off from just using the existing syntax and just collapsing the declare and end keywords with the surrounding block (which I've never actually done):
```
loop declare
   Pos renames Index (Path, "/");
begin
   exit when Pos = 0;
   Delete (Path, Path'First, Pos); -- some processing with position of "/"
end; end loop; 
```
... that's a bit odd (formatting wise), but it's legal and it's sort of similar to avoiding the extra indented block.

The existing syntax isn't bad and has some readability advantages - especially for seeing the scope of relevant variables.

There are definitely times when I feel I'd rather not have to write a declare block, but I'm not sure I'd want to change the language for those occasions; usually it just drives other changes and the code does alright by it.

~~For the proposal, would it be possible to declare nested subprograms without a declare block as well?  It doesn't appear to discuss it, but it seems a relevant question.  The C-inspired languages don't really have those (nested subprograms), but then they have lambda closures...~~

(note: updated for bad code blocks; we all rejected these proposals, you can at least see what I meant, now)

**Comment by CKWG** [comment]
Your examples are unreadable. We're discussing making syntax smaller yet keeping it readable! So what do you want to say?

Your first example (no nice indentation found):
```
declare
  loop Pos renames Index (Path, "/");
begin
  exit when Pos = 0;
  Delete (Path, Path'First, Pos); -- some processing with position of "/"
end loop; 
```
In which respect is this better than the traditional:
```
loop
  declare
    Pos renames Index (Path, "/");
  begin
    exit when Pos = 0;
    Delete (Path, Path'First, Pos); -- some processing with position of "/"
  end;
end loop;
```
This whole discussion about "simplifying" syntax becomes laughable and pathetic just for saving one or two keywords.

Basically you seem to be rejecting the idea but yet try to give examples supporting it, some of them as weird as using objects before declaration (again I can't even find a nice indentation):
```
declare
  while Pos > 0
   loop
 Pos renames Index (Path, "/");
begin
   Delete (Path, Path'First, Pos); -- some processing with position of "/"
end loop;
```

And then finally
> For the proposal, would it be possible to declare nested subprograms without a declare block as well? It doesn't appear to discuss it

Yes, it does! Look at the AdaCore link. I have already lamented this before, perhaps not obviously enough:
>Future possibilities ... allow arbitrary declarations

In my opinion, this destroys Ada syntax.

I'm not going to try making your further dubious code examples more readable.

Please let us stop the nonsence.

**Comment by joshua-c-fletcher** [comment]
regarding my example where I used an object before declaring it - that was certainly no good, and not what I intended, but you're right.

The thought was to keep the concept of declarative regions (with distinct declarative parts) that Ada has had all along, but allow constructs that currently don't have their own optional declarative parts to have them, and thus the 'end' for the existing region would also end the scope of the declared items in the declarative part without needing an additional nested or wrapping declare block.

I had written it with indentation but used the wrong kind of code block in the comment; it was meant to be seen as a "declare loop", with declared items indented after that, otherwise following a pretty normal indentation.
```
declare loop
   Pos renames Index (Path, "/");
begin
   exit when Pos = 0;
   Delete (Path, Path'First, Pos); -- some processing with position of "/"
end loop; 
```
I concede, though, it's not great, and was worse with "declare if" and especially "declare else" blocks.

The final example in my previous post of using the existing syntax, but collapsing the declare and end keywords was supposed to look like this:
```
loop declare
   Pos renames Index (Path, "/");
begin
   exit when Pos = 0;
   Delete (Path, Path'First, Pos); -- some processing with position of "/"
end; end loop; 
```
That's valid Ada, but weird formatting.  I'm content to reject those approaches.

Something that matches exactly what I was going for ends up being identical to what Randy already proposed above, when he wrote:
> Perhaps a better solution (if one is needed at all) would be to allow an optional declarative part on other compound statements.
> the only new thing here is allowing a declarative part on an if statement.
and Blady said he liked it:
```
declare
   Pos renames Index (Path, "/");
loop
   exit when Pos = 0;
   Delete (Path, Path'First, Pos); -- some processing with position of "/"
end loop; 
```
(but for this specific example, it _wouldn't_ be re-evaluating Pos with each iteration, so that loop wouldn't actually work.  As CKWG pointed out, we wouldn't want this construct doing unexpected things like that)

... the re-evaluation each iteration could easily be solved with an expression function instead of a rename:
```
declare
   function Pos return Integer is (Index (Path, "/"));
loop
   exit when Pos = 0;
   Delete (Path, Path'First, Pos); -- some processing with position of "/"
end loop; 
```

Ultimately, I agree; the other suggestions I offered earlier aren't worth it.
I like Randy's version - my initial concern was whether it would be impossible for the parser to distinguish this syntax from a declare block missing its 'end' (that also happens to have a loop or other statement inside)... but the 'begin' is gone, too, so we're saving two lines, and it would be parsable after all.

As for the bigger change of just declaring anything anywhere... we'd lose a lot of structure.

There's defined behaviour for tasks related to the begin, too, for example, and while controlled types don't wait for 'begin' to initialize, the existing syntax provides some added clarity for their scope as well.
There are times when adding a declare block feels like extra overhead, but the structure is part of the character of the language.  Optional declarative parts on other compound statements might be a good addition.

**Comment by joshua-c-fletcher** [comment]
The one 'catch' (if it is that) with having "an optional declarative part on other compound statements" is that it's probably about as common to need a declarative part immediately inside those statements (inside the loop, if, or else, for example)

If you needed a declarative section on the inside, an optional 'declare' part above them doesn't provide that.  For those cases you're still back to a traditional nested declare block (which isn't so bad), or call a subprogram that has its own declarative part.

If it's not so bad to have the declare block inside, maybe we don't need new syntax to avoid it on the outside.

**Comment by CKWG** [comment]
A last word, then I'm going to refrain from answering (my point of view is clear, I''m strongly opposed).

I can live well with the following formatting (why not, it's still valid syntax) in order to save indentation, which seems to be a problem:
```
loop declare
  Pos renames Index (Path, "/");
begin
  exit when Pos = 0;
  Delete (Path, Path'First, Pos); -- some processing with position of "/"
end; end loop;
```

**Comment by joshua-c-fletcher** [comment]
Pondering the ability to name identifiers in case_alternatives in proposal described in issue #34, I wondered if that might be a viable way to provide a solution here.

It could certainly address some occasions where someone might want a renames embedded in a condition, but it doesn't quite cover Blady's original two examples:

Blady's first example:
```
      while (Pos renames Index (Path, "/")) > 0 loop
         Delete (Path, Path'First, Pos); -- some processing with position of "/"
      end loop;
```
Attempt at using the pattern matching feature to solve this:
```
      loop

         case Index (Path, "/") is
            when 0 =>
               exit;
            when Pos : others => -- no identifier in others_alternative!
               Delete (Path, Path'First, Pos); -- some processing with position of "/"
         end case;

      end loop;
```
This doesn't work because naming the identifier is only part of case_alternative, not others_alternative.
... although if we knew a static range for the 'Delete' branch, we could use it and it would work.
It's more lines, but 'Pos' only needs to be evaluated once.


Blady's second example:
```
      if (Count renames Read (FD, Buffer'Address, Buffer'Length)) < Buffer'Length then
         Put_Line
           ("warning only" & Natural'Image (Count) &
            " bytes read."); -- some processing with number of read bytes
      end if;
```
Attempt at using the pattern matching feature to solve this:
```
      case Read (FD, Buffer'Address, Buffer'Length) is

         when Count : 0 .. Buffer'Length - 1 => -- not a static range!

            Put_Line
              ("warning only" & Natural'Image (Count) &
               " bytes read."); -- some processing with number of read bytes

         when others =>

            -- processing when it reads >= than the buffer length.

      end case;
```
This doesn't work because the range isn't static.  But identifiers in case statements are essentially like a renames embedded in conditions - they're just only available when the various 'patterns' they need to match with are static.
